### PR TITLE
Allow empty date field for add command and additional enhancements

### PR DESCRIPTION
**Pull request by stevenlimhw**
- [X] `Date` field is optional for the add command, so default date will be set as the date when the user inputs the `Exercise` (i.e. today's date)
- [X] Refactor `JsonAdaptedTag`
- [X] Add validation of date inputs to reject wrong formats and non-existent dates
- [X] Add tests for `Date`

### CODE REVIEW

**Comment by ee-suan** [comment]
Looks good!

File: src/test/java/gim/model/date/DateTest.java [hunk 1]
```diff
@@ -1,26 +1,63 @@
...
+     * Test for invalid date where the date is non-existent, but is not covered by the regex.
+     * Example: 31st February 2022 is a non-existent date.
+     */
     @Test
-    public void constructor_invalidTagName_throwsIllegalArgumentException() {
-        String invalidTagName = "";
-        assertThrows(IllegalArgumentException.class, () -> new Date(invalidTagName));
+    public void constructor_nonExistentDate_throwsDateTimeParseException() {
+        String nonExistentDate = "31/02/2022";
+        assertThrows(DateTimeParseException.class, () -> new Date(nonExistentDate));
```
Comment (ee-suan) [inline src/test/java/gim/model/date/DateTest.java:54]: Love this detail!

File: src/test/java/gim/model/date/DateTest.java [hunk 2]
```diff
@@ -1,26 +1,63 @@
...
     public void constructor_null_throwsNullPointerException() {
         assertThrows(NullPointerException.class, () -> new Date(null));
     }
 
+    /**
+     * Test for invalid date where the formatting does not conform to dd/MM/uuuu format.
+     */
+    @Test
+    public void constructor_invalidDate_throwsIllegalArgumentException() {
+        String invalidDateEmptyString = "";
```
Comment (ee-suan) [inline src/test/java/gim/model/date/DateTest.java:33]: Maybe test for a string of spaces  "            " as well?

Comment (stevenlimhw) [reply src/test/java/gim/model/date/DateTest.java:33]: The test has been updated!

File: src/test/java/gim/model/date/DateTest.java (hunk 1 above)
Comment (stevenlimhw) [reply src/test/java/gim/model/date/DateTest.java:60]: Thank you!

**Review by ee-suan — APPROVED** [review]

**Pull request merged**